itest/rust/src/native_test.rs

```diff
@@ -0,0 +1,141 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use crate::itest;
+use godot::engine::text_server::Direction;
+use godot::engine::{TextServer, TextServerExtension, TextServerExtensionVirtual};
+use godot::native::*;
+use godot::prelude::*;
+
+use std::cell::Cell;
+
+#[derive(GodotClass)]
+#[class(base=TextServerExtension)]
+pub struct TestTextServer {
+    #[base]
+    base: Base<TextServerExtension>,
+    glyphs: [Glyph; 2],
+    cell: Cell<Option<(Rid, i64)>>,
+}
+
+// Simple function to make these up with one field differing.
+fn sample_glyph(start: i32) -> Glyph {
+    Glyph {
+        start,
+        end: 8,
+        count: 9,
+        repeat: 10,
+        flags: 33,
+        x_off: 999.0,
+        y_off: -999.0,
+        advance: 132.0,
+        font_rid: Rid::new(1024),
+        font_size: 1025,
+        index: 1026,
+    }
+}
+
+#[godot_api]
+impl TextServerExtensionVirtual for TestTextServer {
+    fn init(base: Base<TextServerExtension>) -> Self {
+        TestTextServer {
+            base,
+            glyphs: [sample_glyph(99), sample_glyph(700)],
+            cell: Cell::new(None),
+        }
+    }
+
+    unsafe fn shaped_text_get_carets(&self, shaped: Rid, position: i64, caret: *mut CaretInfo) {
+        // Record the arguments we were called with.
+        self.cell.set(Some((shaped, position)));
+        // Now put something in the out param.
+        *caret = CaretInfo {
+            leading_caret: Rect2::from_components(0.0, 0.0, 0.0, 0.0),
+            trailing_caret: Rect2::from_components(1.0, 1.0, 1.0, 1.0),
+            leading_direction: Direction::DIRECTION_AUTO,
+            trailing_direction: Direction::DIRECTION_LTR,
+        };
+    }
+
+    fn shaped_text_get_glyph_count(&self, _shaped: Rid) -> i64 {
+        self.glyphs.len() as i64
+    }
+
+    unsafe fn shaped_text_get_glyphs(&self, _shaped: Rid) -> *const Glyph {
+        self.glyphs.as_ptr()
+    }
+}
+
+#[itest]
+fn test_native_structures() {
+    // Test construction of a few simple types.
```

Comment:
This tests whether generated structs have a certain layout, right? Maybe make test name a bit more concrete and omit the comment (unless it really adds information).

Reply:
Hmm... testing whether we have a certain layout would be a good idea. Probably an `unsafe` test, but would be useful. To be honest, this test is mainly just checking that the generated files exist and can be imported, which I guess is a bit silly now that I say it out loud.

Reply:
I didn't mean memory layout but "names and types of struct fields".
Which is a good thing to test (we have, but maybe name it `test_codegen` or so...